source/slang/ir-legalize-types.cpp

```diff
@@ -818,6 +818,8 @@ static LegalVal legalizeInst(
             args,
             inst->getOperandCount());
 
+    case kIROp_undefined:
```

Comment:
I am not sure I understand the IR, but this is what is needed to get the test cases to pass.

In the IRFunc, I can see that there is an inst after a param inst that has opcode "kIROp_undefined", with only a non-null type field and 0 operands. I assume that this is intending to keep a reference of the IRType?

However when trying to lower this instruction, we will triger an unimplemented case here. This is somehow not triggered without making the change in this PR, which I don't get why.

Reply:
The `undefined` op is used for the value of an SSA variable before it has been written; if code actually *uses* the `undefined` then it should really be an error, but we need to eliminate dead/useless code before we can go issuing errors like that.

I think the code you have here isn’t quite complete, because a `LegalVal` for a legalized instruction should follow the “shape” of its `LegalType`. That is, if the type of the instruction is a tuple, we should return a tuple of `undefined` values, etc.
 
The main place we see an `undefined` show up today ends up being for empty `struct` variables since the SSA pass thinks they are uninitialized (It never sees a `store` to them). The way you’ve handled `undefined` here works for that one case.

Realistically either lowering or the SSA pass should be smart enough to know that an empty `struct` is always initializer and write a zero argument `makeStruct` to it.

Anyway, your fix here is fine for now, but we should eventually expand it to work with any shape of type.